### PR TITLE
Clear the previous license when an invalid key is applied

### PR DESCRIPTION
**Pull request by zeroliu**
Fixes Brevilabs/obsidian-copilot-private#307

## Why

A user with a working Lifetime key opened Settings, Copilot License, replaced
the stored key with a garbage string, and clicked Apply. Nothing happened. The
badge kept reading `Lifetime`, the "Thanks for being a Copilot Plus user"
welcome modal opened as though the new key had been accepted, no error text
appeared under the field, and every paid gate stayed open for roughly two more
weeks.

Two separate things had to go wrong for that. First, `/license` refuses an
unknown key with HTTP 403 and a reason that names the key's prefix, for example
`Invalid license key (prefix: garbage-ke...)`. The client compared that reason
against the bare string `Invalid license key`, so the prefix suffix broke the
match and an authoritative refusal was reclassified as "server unreachable".
Second, that unreachable path falls back to the entitlement verified earlier in
the session, and nothing tied that entitlement to the license key that earned
it. The refused key inherited the previous key's plan.

The second half is the general one. After any key change, any genuinely unknown
outcome, offline, 5xx, or a gateway error, handed the new key the old key's
access.

## What

| Applying a new key over a working Lifetime key | Before | After |
| --- | --- | --- |
| The API refuses it (403 with its error body) | Badge `Lifetime`, welcome modal, paid gates open | Badge `Inactive`, `Invalid license key` under the field, paid gates closed |
| Server unreachable, 5xx, or an unexplained 403 | Badge `Lifetime`, every paid gate open | `Invalid license key` reported; Plus and self-host gates closed; badge falls back to a generic `Active` and broad paid access persists until the server answers |
| The API accepts it | Badge shows the new plan | Unchanged |
| No key change, server unreachable | Existing plan keeps working offline | Unchanged |

Two things that outage row does not do, both deliberate. It does not name the
previous plan any more, so a refused key can no longer read `Lifetime`. And it
does not revoke anything: the persisted paid flag is not bound to the key that
earned it, and the fix for that unregisters the Plus provider on every valid key
rotation, resetting the user's per-model enablement. Trading a stale badge in an
offline edge case for real state loss on the common path is the wrong trade, so
that half is Brevilabs/obsidian-copilot-private#309.

Revocation happens only when the API itself refuses a key, never on an outage. A
gateway or WAF answering 403 with no API error body is treated as unknown, so a
license server having a bad day costs no paying user their access. A keyless
check no longer reaches the server at all: the endpoint refuses an empty key with
the same 403 it gives a wrong one, and the per-turn and per-model gates call
`validateLicenseKey` without the sign-out `checkIsPaidUser` performs when it finds
no key. The offline window for an unchanged key is untouched.

That 403-with-no-body guard is insurance while sign-out stays destructive, and it
comes out once Brevilabs/obsidian-copilot-private#308 lands.

## Non goal

- Binding the persisted `isPaidUser` flag to its license key, and having
  `useIsSelfHostEligible()` read the key-bound proof rather than re-verifying the
  raw token. Both are Brevilabs/obsidian-copilot-private#309, which is itself
  blocked on Brevilabs/obsidian-copilot-private#308: sign-out deletes the
  ConfiguredModel rows that enablement is keyed on, so clearing the paid flag on
  a key change would reset the user's model curation on every valid rotation.
- The chat error path in `src/utils.ts` matches the same reason text the same
  brittle way, so a lapsed key still shows a raw error there instead of the
  friendly upgrade message. That is pre-existing and unchanged here.
- The expired-license modal now opens for a mistyped key, since that branch
  became reachable. Its "Renew Now" framing is not retuned in this PR.
- No new offline grace behavior, and no change to how any license state renders.

## Screenshot

Not applicable. This PR changes which license state the settings section
reaches, not how any state renders. The diff touches four files, all of them
`src/plusUtils.*` and `src/LLMProviders/brevilabsClient.*`, and no UI component,
copy string, or stylesheet. The reachable states and their exact copy are in the
`What` table above; the `Inactive` badge, the generic `Active` badge, and the
`Invalid license key` message all already exist and are unmodified.

## Risk

**High**

| Criterion | Status | Reason |
| --- | :---: | --- |
| No behavior change, or a cosmetic/copy/docs/config change visible where it renders, or deterministic tests cover the changed behavior | ✅ | Added tests cover the API refusal, an unexplained 403, a 502, an empty stored key, the key-swap fallback, the badge, and the in-flight key swap |
| A defect would fail CI or be obvious on first use | ✅ | Those tests run in CI, and a wrong verdict shows on the next Apply |
| A revert fully restores prior state, including persisted data | ✅ | The key binding is in-memory only, no settings schema change and no migration; a revert needs no data change |
| No auth, permissions, secrets, or input-handling surface changes | ❌ | This is the license entitlement gate: it decides what unlocks paid features |
| No public API, plugin API, message, or on-disk contract changes | ✅ | `BrevilabsApiResult` and `VerifiedEntitlement` are module-private and no exported signature changed |
| No core-path concurrency, async-lifecycle, or state-machine changes | ❌ | `applyEntitlement` now reads the license key before its verification await rather than after |
| No hot-path behavior lacks deterministic coverage | ✅ | Every changed branch has a test that was confirmed to fail against the previous code |
| No new dependency | ✅ | Dependency manifests are unchanged |
| Human-only behavior stays in one feature area and surfaces quickly | ✅ | Confined to license validation, and a wrong result appears on the next Apply or startup check |

Review: inspect `validateLicenseKey` and `parseBrevilabsResponse` in
`src/LLMProviders/brevilabsClient.ts`, then `hasLiveEntitlement`,
`applyEntitlement`, and `verifyCachedEntitlement` in `src/plusUtils.ts`, line by
line. The condition worth the most attention is the one deciding when
`turnOffPaid` runs, since that is the only path that takes access away. Then run
Verification steps 2 to 7, especially step 6, which proves an outage on an
unchanged key costs nothing.

## Verification

1. Build the branch and load it in a vault that has a working paid license key
   applied. Open Settings, Copilot License, and note the plan badge.
2. Replace the key with any garbage string and click Apply. The badge should
   turn to `Inactive`, `Invalid license key` should appear under the field, and
   paid models should stop being offered.
3. Reopen the settings tab. It should still read `Inactive` rather than
   returning to the previous plan.
4. Paste the real key back and click Apply. The plan badge should return and
   paid models should work again.
5. With the real key applied, disconnect from the network, replace the key with
   a garbage string, and click Apply. It should report the key as invalid and
   stop offering Plus models, rather than showing the previous plan name.
6. With the real key applied, disconnect from the network and click Apply without
   changing the key. Access should be unaffected, which is what proves an outage
   costs a paying user nothing.
7. Reconnect, paste the real key, and click Apply. Access should come back.
8. Clear the key field entirely, then start a turn that @-mentions two agents.
   It should refuse the fan-out without opening the expired-license modal, since
   having no key is not a refusal to revoke on.

## Note

The endpoint contract was confirmed against the live server before the fix:
`POST /v1/license` with an unknown key returns HTTP 403 with
`detail.error = "FORBIDDEN"` and a `reason` carrying the key prefix. The
existing unit test passed throughout the bug because it stubbed the bare literal
`Invalid license key`, a shape the server no longer sends.

The presence of that `detail` body, not the 403 alone, is what authorizes
revocation. Infrastructure returns 403 with an HTML page and no API error body,
and reading that as a verdict would revoke every paying user who happened to
check in during an outage.

The same probe showed an empty `license_key` returns the identical 403 and body
as a wrong one, while omitting the field returns 400. Since the plugin always
sends the field, a keyless check was indistinguishable from a refusal, which is
why `validateLicenseKey` now answers locally when no key is stored.

### CODE REVIEW

File: src/plusUtils.ts [hunk 1]
```diff
@@ -157,12 +166,17 @@ let verified: VerifiedEntitlement = NO_VERIFIED_ENTITLEMENT;
...
+ * belong to the token AND the license key settings currently hold, and the
+ * signed `exp` must not have passed. This is the offline window — signature-backed,
+ * so unlike the persisted `isPaidUser` an edited `data.json` cannot widen it.
  */
 function hasLiveEntitlement(): boolean {
-  return verified.token === getSettings().entitlementToken && Date.now() < verified.expiresAt;
+  const settings = getSettings();
+  return (
+    verified.token === settings.entitlementToken &&
+    verified.licenseKey === settings.plusLicenseKey &&
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/plusUtils.ts:177]: Confirmed and fixed the overclaim, but not taking the proposed fix. Verified the reported state with a probe test rather than reasoning about it:

`useLicenseState()` returns `{"status":"active"}` and `isPaidEnabled()` returns `true` after a key swap with no server verdict. You are also right that my swapped-key test masked it by presetting `isPaidUser: false`. Measured values for that scenario:

| Reader | Result |
| --- | --- |
| badge | `Active` (generic, not the previous plan name) |
| `isPaidEnabled()` | `true` |
| `isPlusEnabled()` / `canUseMultiAgent()` | `false` |
| `isSelfHostModeValid()` | `false` |

The PR body claimed this row closed every paid gate. That was wrong and is now corrected to the measured behavior.

Not invalidating the persisted paid state on key change here. Clearing `isPaidUser` flips `plusSyncNeeded` (`src/modelManagement/setup/copilotPlusSync.ts:138`), which calls `unregisterPlusProvider` and cascade-removes the provider row, its configured models, and its backend refs. `registerPlusProvider` then re-enrolls only `autoEnrollModelIds`, the default-on subset, so every routine rotation to a **valid** key would silently reset the user's per-model enablement. That trades a stale badge in an offline edge case for real state loss on the common path.

Doing it without that cost means persisting which key earned the flag, which is a settings field plus a migration for already-written data. That is logancyang/obsidian-copilot-preview#353, with the measurements and both options.

File: src/plusUtils.ts [hunk 2]
```diff
@@ -564,12 +578,17 @@ export function turnOffPaid(app?: App): void {
...
+  // otherwise tag this proof with the new key while it carries the old key's
+  // claims, which is the inheritance this binding exists to prevent.
+  const { userId, plusLicenseKey } = getSettings();
+  const claims = await verifyEntitlement(token, { expectedUserId: userId });
   if (!claims) {
     return false;
   }
   verified = {
     token,
+    licenseKey: plusLicenseKey,
```
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/plusUtils.ts:591]: Not reachable, so not adding the branch. The await this describes is `verifyEntitlement`, which is two WebCrypto calls (`subtle.importKey` then `subtle.verify`) against public keys embedded in `src/entitlement/publicKeys.ts`. No network, no I/O, no JWKS fetch.

The overlap the comment needs is: key A's `/license` response arrives, passes the key-changed guard, enters `applyEntitlement`, and then key B's entire Apply plus network round trip plus 403 plus `turnOffPaid` all complete before key A's CPU-bound verify resolves. A full HTTP round trip cannot finish inside a sub-millisecond crypto call.

The commit already reads the key before the await rather than after, which is correct sequencing and costs nothing, and there is a test for it. Adding a post-await bail on top would also need `validateLicenseKey` to re-check before `markPaidPendingEntitlement`, since returning `false` from `applyEntitlement` currently routes into it and would set `isPaidUser: true` for the new key. That is two branches guarding a state that cannot occur, which this repo ranks as a P1 finding in its own right.

File: src/plusUtils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (zeroliu) [reply src/plusUtils.ts:177]: Agreed, and it is the same root as the persisted-flag finding: this PR key-bound the in-memory proof, while `useIsSelfHostEligible()` re-verifies `settings.entitlementToken` directly and never consults the key.

Confirming the halves that do hold today, since they bound the impact. The runtime gate is closed: `isSelfHostModeValid()` reads `hasVerifiedFeature("self_host")`, which now fails the key comparison, so a swapped key cannot actually route searches to self-host. What survives is the settings-tab reading and the toggle it leaves enabled.

Worth noting the hook must keep treating an unverifiable token as unknown rather than as a downgrade. It only clears `enableSelfHostMode` when a token verifies and lacks the feature, because clearing on a failed verify would destroy a preference no later success restores. A key mismatch is a different signal from a failed verify and should reject without touching the preference.

Tracked with the persisted-flag work in logancyang/obsidian-copilot-preview#353, since consuming the key-bound proof is the same change on both readers.

[the remark this answers is not in the document]
Comment (zeroliu) [reply src/plusUtils.ts:177]: Followed the root cause further: filed logancyang/obsidian-copilot-preview#354.

Sign-out is destructive, and that is what blocks the fix you proposed. `ModelManagementCoordinator.removeProvider` deletes the ConfiguredModel rows, which destroys the `configuredModelId`s that `BackendConfig.enabledModels` is keyed on, so re-registration mints new ids and auto-enroll turns on only the default subset. Re-registering over a surviving provider is already lossless (`#reconcileModels` writes only real deltas), which is why rotating a key while `isPaidUser` stays true loses nothing today.

Once sign-out deactivates rather than deletes, clearing `isPaidUser` on key change costs nothing and your finding becomes a small, correct change. #354 carries the read-path survey: two projection sites, four consumers, and `EnabledBackendEntry` is already a discriminated union that `chatModelSelection.ts` narrows, so the compiler names every site that has to decide.

[the remark this answers is not in the document]
Comment (zeroliu) [reply src/plusUtils.ts:177]: Housekeeping: the issues these threads reference moved. I had filed them in `logancyang/obsidian-copilot-preview`, which the routing map removed as the legacy tracker. Re-filed in the correct tracker and the source comments and test names in the diff now point there:

- Brevilabs/obsidian-copilot-private#307, this PR
- Brevilabs/obsidian-copilot-private#309, the persisted paid flag, this thread
- Brevilabs/obsidian-copilot-private#308, non-destructive sign-out, which #309 is blocked on
